design-proposals/memory-hotplug.md

```diff
@@ -0,0 +1,209 @@
+
+## User Stories
+
+As a VM user, I would like to dynamically increase or decrease the number of Memory used by a VM without requiring the
+VM to restart.
+
+## Repos
+
+kubevirt/kubevirt
+
+## Design
+
+In VM objects, the existing Memory `guest` field will now be dynamic, provided that it has been declared as such.
+Declaring the field as dynamic will be done by adding a `memory` entry to a section under the VM spec,
+called `liveUpdateFeatures` (definition during CPU hotplug implementation).   
+Please note that prior to defining this field the hotplug action would not take place i.e. changing the number of guest
+will be staged until further reboot. Under the `memory` entry, it will be possible to define a maximum number of guest.
+That number is needed by LibVirt and will default to 4 times the initial number of sockets if not set. That default
+value will be configurable in the KubeVirt CR. On VM startup, that value will translate to a new `maxGuest` entry
+under `spec.domain.memory` in the VMI object. However, all VMI Memory fields will stay immutable, and Memory guest will
+only be adjustable on the VM object. Once the hoplug action begins, subsequent changes to the number of guest would be
+rejected until the hotplug process completes. Memory Resource update on the pod level must occur before the actual
+hotplug action on the Libvirt level, hence, during the hotplug action a notifitcation will be created upon successful
+update of the Memory resources on the pod level. It will indicate that now it safe to proceed to the hotplug action on
+the Libvirt/QEMU level.
+
+## Drawbacks / Linmitations
+
+* Workload update method should be defined explicitly as `VMLiveUpdateFeatures` under the Kubevirt configuration.
+* This feature will be incompatible with memory requests/limits, at least initially, since those values need to
+  dynamically scale according to the current number of enabled memory
+* Set the memory increment (new_guest - old_guest) for online memory expansion An integer multiple of 128 MB (operating
+  system requirements)
+* Memory hotplug depends on numa, VMs will have to be `NUMAable` to enable Memory hotplug
+
+#### API Examples
+
+##### VM Spec API
+
+```yaml
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+metadata:
+  labels:
+    kubevirt.io/vm: vm-cirros
+  name: vm-cirros
+spec:
+  running: false
+  liveUpdateFeatures:
+  +   memory
++     maximumGuest: 4Gi
+  template:
+    metadata:
+      labels:
+        kubevirt.io/vm: vm-cirros
+    spec:
+      domain:
+        memory:
+          guest: 4Gi
+```
+
+##### VMI Spec API
+
+```yaml
+apiVersion: kubevirt.io/v1
+kind: VirtualMachineInstance
+metadata:
+  labels:
+    special: vm-cirros
+  name: vm-cirros
+spec:
+  domain:
+    memory:
++     maxGuest: 10Gi
+  guest: 8Gi
+
+```
+
+## Status API
+
+The VMI status must indicate whether Memory hotplug is taking place. As the hotplug process begins the current state
```

Comment:
Can you reword this?

The VM/VMI will have a new field `status.memory` that looks like the following:

```yaml
status:
  memory:
    guestAtBoot: 4Gi
    guestCurrent: 8Gi
```

It's pretty self-explanatory but `guestAtBoot` shows how much memory the VM booted with and `guestCurrent` shows the currently available memory to the guest, so of course this last field changes when a hotplug is completed.

Reply:
I don't think you need to record what you want to change, you just need to have the original value。 `guestCurrent` is not necessary and can be represented by `spec.domain.memory.guest`

Reply:
We do actually need `guestCurrent` as `spec.domain.memory.guest` can be different that `status.memory.guestCurrent` after a successful hot(un)plug operation as you're not always guaranteed to get all the memory you requested.

This is especially true when you're resizing down memory.

`guestCurrent` track how actual memory you have, `guest` track what the request was.

Reply:
Is your guestCurrent always equal to `spec.domain.memory.maxGuest` - `spec.domain.memory.guest` ?  If so, if this value is useful elsewhere, it can be calculated directly, and there is no need to add a field

Reply:
After reading the https://github.com/kubevirt/kubevirt/pull/10184 code, I understand what you mean. What I want to express is whether it can be implemented like cpu without caring about whether the operation is successful.   I think you only need to have one field in `status.memory` to be consistent with the domain, in fact, it is not very meaningful to record the actual memory size of the vm.  Can you give me an example if you feel it's important?

Reply:
It is implemented like that, in https://github.com/kubevirt/kubevirt/pull/10184 if you request a memory hotplug KubeVirt will mark it as successful if you can hot(un)plug any memory at all, but we have to note somewhere what's the guest available memory as it really can be different from what requested, especially when resizing down memory.

Otherwise how can you tell how much memory your guest currently has? 

A simple example, you have a 4Gi VM, you want to hotplug another 4Gi of memory, for some reason the system can only give you 2 so you end up with a 6Gi VM and the hotplug is marked as successful. If you don't write in `status.memory.guestCurrent` how can you tell how much memory you have? That's a pretty important thing to know.

Reply:
Partial success when the requested resources are insufficient?